pragma-sdk/pragma_sdk/common/fetchers/fetchers/dexscreener.py

```diff
@@ -0,0 +1,148 @@
+import time
+
+from typing import List
+from aiohttp import ClientSession
+
+from pragma_sdk.common.fetchers.handlers.hop_handler import HopHandler
+from pragma_sdk.common.types.entry import Entry, SpotEntry
+from pragma_sdk.common.types.pair import Pair
+from pragma_sdk.common.types.currency import Currency
+from pragma_sdk.common.exceptions import PublisherFetchError
+from pragma_sdk.common.fetchers.interface import FetcherInterfaceT
+from pragma_sdk.common.logging import get_pragma_sdk_logger
+
+from pragma_sdk.onchain.client import PragmaOnChainClient
+
+logger = get_pragma_sdk_logger()
+
+
+class DexscreenerFetcher(FetcherInterfaceT):
+    """
+    Dexscreener fetcher.
+    NOTE: Only works for Starknet at the moment.
+
+    Also, the API is still in beta so we expect breaking changes to happen.
+    """
+
+    publisher: str
+    pairs: List[Pair]
+    _client: PragmaOnChainClient
+
+    hop_handler = HopHandler(
+        hopped_currencies={
+            "USD": "USDC",
+        }
+    )
+    SOURCE = "DEXSCREENER"
+    BASE_URL: str = "https://api.dexscreener.com/latest/dex"
+
+    # NOTE: We only check for starknet for now
+    CHAIN_ID = "starknet"
+
+    async def fetch(
+        self, session: ClientSession
+    ) -> List[Entry | PublisherFetchError | BaseException]:
+        """
+        Fetch prices from all pairs from Dexscreener.
+        """
+        entries = []
+        for pair in self.pairs:
+            entries.append(self.fetch_pair(pair, session))
+        return list(await asyncio.gather(*entries, return_exceptions=True))  # type: ignore[call-overload]
+
+    async def fetch_pair(
+        self, pair: Pair, session: ClientSession
+    ) -> SpotEntry | PublisherFetchError:
+        """
+        Fetch the price for a pair and return the SpotEntry.
+
+        NOTE: The currencies of the pair must have a starknet_address.
+        """
+        hopped_pair = self.hop_handler.get_hop_pair(pair) or pair
+        if hopped_pair.base_currency.starknet_address == 0:
+            return PublisherFetchError(
+                f"Failed to fetch data for {hopped_pair} from Dexscreener: "
+                f"{hopped_pair.base_currency.id} starknet_address is None."
+            )
+        if hopped_pair.base_currency.starknet_address == 0:
+            return PublisherFetchError(
+                f"Failed to fetch data for {hopped_pair} from Dexscreener: "
+                f"{hopped_pair.quote_currency.id} starknet_address is None."
+            )
+        return await self._fetch_dexscreener_price(pair, session)
+
+    async def _fetch_dexscreener_price(
+        self,
+        pair: Pair,
+        session: ClientSession,
+    ) -> SpotEntry | PublisherFetchError:
+        """
+        Query the dexscreener API and construct the SpotEntry.
+
+        NOTE: It is really unclear at the moment how the pair is actually constructed,
+        sometimes the quote asset is in front of the base asset...
+        To be sure it works, we try both.
```

Comment:
but if we try both then the price won't be the same, like ETH/USDC != USDC/ETH

Reply:
I agree, but for example for EKUBO the pair EKUBO/USDC as addresses is formatted the other way around and it returns the EKUBO/USDC price

Reply:
just checked, weird but true indeed